### DISCUSSION TITLE
Foot controller

### DISCUSSION THREAD

**Discussion by StaffanMelin** [cardonabits/haxo-rs#44] [Ideas]
So I am thinking of making a foot controller for the Haxo.

initial thoughts and ideas
* expression pedal for controlling brightness (filter cutoff)
* configurable
* add 4 buttons for switching sounds etc
* display? LEDs!

Making it
* adding a small MCU with Analog in to the Haxo (useful for other stuff too, ie interfacing): I am thinking about a XIAO (https://wiki.seeedstudio.com/Seeeduino-XIAO/) as it is super small, versatile, and I have one already
* pressure sensing: resistive film? pot with a return spring? bend resistor?
* put the MCU in the foot controller instead? Better to send digital than analog thru a wire to the RPi on the Haxo?
* use piezos for foot switches as they are prob more durable...?

Any thoughts?

**Comment by jcard0na** [comment]
Hi @StaffanMelin,

This sounds like a fun project.  I would recommend also opening a discussion in https://github.com/cardonabits/haxo-hw/discussions where you might get feedback from more hardware-oriented folks.

My recommendation on hardware selection would be that it is easy to source.  If you want people to contribute, the hardware should be cheap and widely available.  Also, if you stay in Rust you will not introduce language barriers to those already familiar with `haxo-rs`.

So why not stay with a Raspberry Pi Zero?  They are cheap, widely available, familiar to most haxophonists and small enough to fit on a foot pedal.  In fact there are several guitar foot pedals that use it:  [1](https://www.electrosmash.com/pedal-pi) [2](https://dylanbeattie.net/2020/05/17/turning-a-raspberry-pi-zero-into-a-usb-footpedal.html) ... or even the RPI4 [3](https://github.com/GuitarML/NeuralPi)

Looking forward to see what you come up with!

**Reply by StaffanMelin** [reply]
Thank you for your thoughts.

For this project I think an MCU as apposed to a SBC will be more suitable. I don't need (nor want) an OS and to fiddle with SD cards etc, and I want the built in ADCs of a MCU. So anything from an Arduino to a RP2040 sounds about right.

As well as being really cheap, the XIAO has 12 bit ADC resolution (and 11 GPIOs), and even a 10 bit DAC (CV out, right, for connecting to modular stuff, even though it is only 3.3V).

We'll see where I end up!

**Comment by jcard0na** [comment]
I see.  In that case I would just ask you to consider an option that be programmed in Rust.  I've been programming STM32s in Rust for some time now, and never missed my C/C++ days... 
The STM32s have tons of variants, so you should be able to find the built-in peripherals you want for your project (https://www.st.com/content/st_com/en/stm32-mcu-product-selector.html).

**Reply by StaffanMelin** [reply]
I've never really understood what STMs are? :) Just a brand, licensing the ARM Cortex processors? But with good SW/lib support?

I've worked a lot with the Daisy Seed, a Cortex M7 with an audio codec, very fast for doing DSP: https://electro-smith.com/products/daisy-seed. There seems to be STM stuff in the toolchain...

But for the foot controller application, I don't need any more processing power than what an Arduino provides. No audio processing, that will all run on the RPi.

The Seeed XIAO SAMD21 is cheap with great io. It has a Cortex M0.

Rust is certainly an option, I'll promise to think about it! :)

**Reply by jcard0na** [reply]
Thank you!

**Comment by StaffanMelin** [comment]
So the prototype hardware is finished as a simple ... prototype! :)

![hexo](https://github.com/cardonabits/haxo-rs/assets/33361311/3a904856-ed13-4113-90b2-a9638d594602)

It is powered by the inexpensive yet powerful Seeed Xiao (SAMD21 version), chosen because a) it has good connectivity b) I already had one. :)

I also had some pieces laying about from a previous controller project so I simply put them together and connected them up to the Xiao. Well, I had to replace the two pots on the mod and pitch wheel (that I salvaged from a broken keyboard in my common garbage room).

1. Xiao
2. Mod wheel, for LFO intensity etc
3. Pitch bend wheel (with a spring so it centers)
4. Distanse sensor (ultrasonic), it can measure the distanse from eg a foot with good precision. Filter cutoff?
5. Not connected yet, but it is a 4-channel DAC with 12 bits. I am going to use it to connect my haxo to my analog gear (pitch, control x 3), I will also add a gate out directly from the Xiao.
6.  USB connection. I am going to connect the Xiao via USB (Serial) to the Haxo. I already have some C++ test code running on the Xiao and the laptop (which is going to be replaced by the Raspberry Pi connected to the Haxo).

Future plans include connecting a small OLED over I2C, as well as a simple analog button combo, so I can build a simple interface on the Xiao. Then the ports of the Xiao are all but taken.

All in all, pretty standard stuff, readily available.

I hope to have a working example this weekend. If I find it useful, I will add documentation and code (I will anyway! :) ).

I hope to be able to use the 2, 3 and 4 with my feet, we'll see. Whatever, the important thing is having some kind of analog input and get it to the Raspberry Pi on the haxo.

I call it the "Hexo"! :)

**Reply by Ekolide** [reply]
That's cool! How do you interface the Haxophone with the Xiao?

**Reply by StaffanMelin** [reply]
Thank you! :) I am using the USB-port on the XIAO and one on the Raspberry Pi (I am using a 4B so I have plenty). So it is serial (se point 6 above). I have a really simple protocol, sending comma separated integers, one for each "sub-controller" (wheel, ultrasonic sensor etc). Working fine!

I did some coding this weekend, building a sort of modulation matrix between the controller and the haxo. I hope to finish a test this week. :)

**Comment by stardusthamster** [comment]
Hi @StaffanMelin ,

thank you very much for sharing your project!

I am interested to do some sort of remote control like your foot pedal as well. My idea is much simpler. I have a Seeed Xiao left over.
So what I would like to do is the following:
- connect the xiao to some buttons and some LEDs
- each button has a corresponding LED. So when button 1 is pushed, LED 1 will light on etc
- I want to use the buttons to switch several haxo settings at once. For example: 
  * button 1: switch sound to flute and change transpose mode to c
  * button 2: switch sound to clarinet and change transpose mode to b
  * button 3: switch back to default tenor sax and transpose mode b

So my questions about your project are the following:

1. the arduino code is this one: https://github.com/StaffanMelin/haxo-cs/blob/master/example-haxo-hexo/arduino/hexo/hexo.ino As I understand you write 3 comma separated values to the serial port. 
Is this the case? 
So I could use this to write so serial which button has been pushed (or something similar). And I could read from serial the response sent by the Raspberry pi?

2. But how does it work on the Raspberry Pi side? I found the testfile https://github.com/StaffanMelin/haxo-cs/blob/master/example-haxo-hexo/computer_test_read/test.cpp and I understand that it reads from the serial port.
Where is the processing done? Which other files from haxo-hexo would I need to use?
 
 Thank you very much in advance!

**Reply by StaffanMelin** [reply]
Hi @stardusthamster and thank you for your interest in my work!

1. In short, yes! With push buttons, the Arduino code will be even simpler! (The three-numbers protocol is just something I made up, I like when things are human readable and I can use the Serial Monitor in the Arduino IDE to test it.)

2. I have some more code on my computer, but things are moving slowly due to lack of time (or focus :) ). I will try to wrap things up this weekend.

Note that the code on the RPi is very much integrated with my sound generating stuff, the dstudio-audio/DSP-library. For you to use this with the original Haxophone firmware you have to add stuff in Rust. A more general solution would be to add MIDI IN to the Haxophone firmware, and send MIDI from the foot controller (over USB). Hmmm, why didn't I think of that in the first place?

This is just a short answer, I'll write more this weekend! Once again, thank you for your interest and comment!

**Reply by stardusthamster** [reply]
Thank you for your fast answer!

I found the following tutorial that explains how to connect the Raspi Pi can be connected to an Arduino. The final example shows how to push a button on Arduino and get some response from the Raspi Pi and switch on an LED to show the response.
[(https://roboticsbackend.com/raspberry-pi-arduino-serial-communication/#What_is_Serial_communication_with_UART)]

Will have a look at this one next week, I just ordered some parts to get me really started...

Looking forward to your details!

**Reply by StaffanMelin** [reply]
Unfortunately I didn't have the time to write anything up  this weekend, but work is moving forward, so hopefully this week will see some progress. I will at least push a lot of new code to the repo so you can see how to handle the communication.

I just want to stress that my code is meant to work with my sound library [rtDStudio](https://github.com/StaffanMelin/haxo-cs/tree/master/rtDStudio), not the original firmware written in Rust. I only work in C++. rtDStudio is a rtAudio-version of my Open Frameworks addon (https://github.com/StaffanMelin/ofxDStudio), that is running on the Raspberry Pi instead of the original Rust code. For me, this gives me more possibilities, but it is more of a electronic music library, than a saxophone emulator. :)

**Reply by StaffanMelin** [reply]
Sorry @stardusthamster, nothing new to push right now, as I am rewriting some basic DSP stuff that my haxo-cs depends on. This is to allow switching between different patches that require different synth-engines.

But if you look at https://github.com/StaffanMelin/haxo-cs/blob/master/rtDStudio/src/dhaxo.cpp you can see how the serial data is received and handled.

**Reply by stardusthamster** [reply]
Thank you for the link to dhaxo.cpp. 

I think, this will help me a lot when I reach the point where I want to integrate the stuff into the haxophone software. But this will take some more time...

**Comment by stardusthamster** [comment]
Hi  @StaffanMelin,

I pushed my first steps here:

https://github.com/stardusthamster/haxo-misc/tree/main/misc

So if you want to take a look. But it is still work in progress...

**Comment by StaffanMelin** [comment]
How is it going?

I don't know Rust unfortunately, but it looks a bit like C++. One day I 
have to learn it, it seems to get a lot of traction.

Anyway, good luck!

Staffan


2024-08-16 16:57 skrev stardusthamster:

**Reply by stardusthamster** [reply]
Thank you for your interest!

I just succeeded to connect the arduino via serial to the haxophones three pin connector and started a read/write from serial program on both and they talked to each other!

So this is a major step forward for my project. I will publish the arduino code soon. On haxophone I used the pi_serial_test program that is already in my repo.

**Reply by stardusthamster** [reply]
Hi Staffan,
I just did another step and wrote a script to change instuments. So another major step is done!

**Reply by StaffanMelin** [reply]
Great! Inspiring! Keep up the good work, @stardusthamster!
